src/foam/dao/index/TreeIndex.java

```diff
@@ -84,8 +125,20 @@ public SelectPlan planSelect(Object state, Sink sink, long skip, long limit, Com
       return NotFoundPlan.instance();
     }
 
-//    Binary expr = isExprMatch(predicate, In.class);
-//    Binary expr = isExprMatch(predicate, Eq.class);
+
+    Binary expr;
+    Object subTree = state;
+
+    //expr = isExprMatch(predicate, In.class);
+    //if ( expr != null && Math.log(((TreeNode)state).size)/Math.log(2) * ((In) predicate) < ((TreeNode) state).size )
+    expr = isExprMatch(predicate, Eq.class);
+    if ( expr != null ) {
+      subTree = ((TreeNode) state).get((TreeNode) state, expr.getArg2().f(expr), prop_);
+      if ( subTree == null ) return new NotFoundPlan();
+      if ( ! (subTree instanceof TreeNode) )
```

Comment:
Why is this check necessary? Why do you ever need to know what kind of subTree it is? You can't know all of the possibilities.